crates/ruma-html/src/html/matrix.rs

```diff
@@ -0,0 +1,721 @@
+                            continue;
+                        }
+
+                        let sub_len = (language_end - language_start) as u32;
+                        data.language = Some(attr.value.subtendril(language_start as u32, sub_len));
+
+                        if match_start != 0 || language_end != value_str.len() {
+                            // There are other classes, keep the whole attribute for the conversion
+                            // to be lossless.
+                            remaining_attrs.insert(attr.clone());
+                        }
+
+                        break;
+                    }
+
+                    if data.language.is_none() {
+                        // We didn't find the class we want, keep the whole attribute.
+                        remaining_attrs.insert(attr.clone());
+                    }
+                }
+                _ => {
+                    remaining_attrs.insert(attr.clone());
+                }
+            }
+        }
+
+        (data, remaining_attrs)
+    }
+}
+
+/// The supported data of a `<span>` HTML element.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct SpanData {
+    /// `data-mx-bg-color`, the background color of the text.
+    pub bg_color: Option<StrTendril>,
+
+    /// `data-mx-color`, the foreground color of the text.
+    pub color: Option<StrTendril>,
+
+    /// `data-mx-spoiler`, a Matrix [spoiler message].
+    ///
+    /// The value is the reason of the spoiler. If the string is empty, this is a spoiler
+    /// without a reason.
+    ///
+    /// [spoiler message]: https://spec.matrix.org/latest/client-server-api/#spoiler-messages
+    pub spoiler: Option<StrTendril>,
+}
+
+impl SpanData {
+    /// Construct an empty `SpanData`.
+    fn new() -> Self {
+        Self { bg_color: None, color: None, spoiler: None }
+    }
+
+    /// Parse the given attributes to construct a new `SpanData`.
+    ///
+    /// Returns a tuple containing the constructed data and the remaining unsupported attributes.
+    #[allow(clippy::mutable_key_type)]
+    fn parse(attrs: &BTreeSet<Attribute>) -> (Self, BTreeSet<Attribute>) {
+        let mut data = Self::new();
+        let mut remaining_attrs = BTreeSet::new();
+
+        for attr in attrs {
+            if attr.name.ns != ns!() {
+                remaining_attrs.insert(attr.clone());
+                continue;
+            }
+
+            match attr.name.local.as_bytes() {
+                b"data-mx-bg-color" if data.bg_color.is_none() => {
+                    data.bg_color = Some(attr.value.clone());
+                }
+                b"data-mx-color" if data.color.is_none() => data.color = Some(attr.value.clone()),
+                b"data-mx-spoiler" if data.spoiler.is_none() => {
+                    data.spoiler = Some(attr.value.clone());
+                }
+                _ => {
+                    remaining_attrs.insert(attr.clone());
+                }
+            }
```

Comment:
The `if … is_one()` forbids to overwrite a value with multiple attributes. Is it the expected behaviour defined by the spec?

Reply:
The spec doesn't say anything about this.

Maybe the safeguards could be removed here. When writing the tests I noticed that html5ever seems to de-duplicate attributes so we only ever get the first one.

Reply:
I removed the guards